File: Robotics_Kinematics_and_Dynamics_Course/Chapter16/Lesson3/Chapter16_Lesson3.java
import org.apache.commons.math3.linear.MatrixUtils;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.SingularValueDecomposition;

public class ParallelSingularity {

    public static class SingularityInfo {
        public String type;
        public double detA;
        public double detB;
        public int rankA;
        public int rankB;
        public double kappaA;
        public double kappaB;
    }

    public static SingularityInfo classify(double[][] Adata,
                                           double[][] Bdata,
                                           double tol) {
        SingularityInfo info = new SingularityInfo();

        RealMatrix A = MatrixUtils.createRealMatrix(Adata);
        RealMatrix B = MatrixUtils.createRealMatrix(Bdata);

        info.detA = new org.apache.commons.math3.linear.LUDecomposition(A).getDeterminant();
        info.detB = new org.apache.commons.math3.linear.LUDecomposition(B).getDeterminant();

        SingularValueDecomposition svdA = new SingularValueDecomposition(A);
        SingularValueDecomposition svdB = new SingularValueDecomposition(B);

        double[] sA = svdA.getSingularValues();
        double[] sB = svdB.getSingularValues();

        info.rankA = 0;
        info.rankB = 0;
        for (double v : sA) {
            if (v > tol) info.rankA++;
        }
        for (double v : sB) {
            if (v > tol) info.rankB++;
        }

        info.kappaA = (sA[sA.length - 1] > tol)
                      ? sA[0] / sA[sA.length - 1]
                      : Double.POSITIVE_INFINITY;
        info.kappaB = (sB[sB.length - 1] > tol)
                      ? sB[0] / sB[sB.length - 1]
                      : Double.POSITIVE_INFINITY;

        int m = A.getRowDimension();
        boolean typeI  = info.rankB < m;
        boolean typeII = info.rankA < m;

        if (typeI && typeII) {
            info.type = "Type III (combined)";
        } else if (typeI) {
            info.type = "Type I (serial)";
        } else if (typeII) {
            info.type = "Type II (parallel)";
        } else {
            info.type = "Regular";
        }

        return info;
    }

    public static void main(String[] args) {
        double[][] A = { {1, 0, 0}, {0, 1, 0}, {0, 0, 1} };
        double[][] B = { {1, 0, 0}, {0, 1, 0}, {0, 0, 1} };

        SingularityInfo info = classify(A, B, 1e-6);
        System.out.println("Type: " + info.type);
        System.out.println("det(A) = " + info.detA + ", det(B) = " + info.detB);
        System.out.println("rank(A) = " + info.rankA + ", rank(B) = " + info.rankB);
        System.out.println("kappa(A) = " + info.kappaA + ", kappa(B) = " + info.kappaB);
    }
}
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter16/Lesson3/Chapter16_Lesson3.cpp
#include <iostream>
#include <Eigen/Dense>

using Eigen::Matrix3d;
using Eigen::JacobiSVD;

struct SingularityInfo {
    std::string type;
    double detA;
    double detB;
    int rankA;
    int rankB;
    double kappaA;
    double kappaB;
};

SingularityInfo classifySingularity(const Matrix3d& A,
                                    const Matrix3d& B,
                                    double tol = 1e-6)
{
    SingularityInfo info;

    info.detA = A.determinant();
    info.detB = B.determinant();

    JacobiSVD<Matrix3d> svdA(A);
    JacobiSVD<Matrix3d> svdB(B);

    Eigen::Vector3d sA = svdA.singularValues();
    Eigen::Vector3d sB = svdB.singularValues();

    info.rankA = (sA.array() > tol).count();
    info.rankB = (sB.array() > tol).count();

    info.kappaA = (sA(2) > tol) ? sA(0) / sA(2) : std::numeric_limits<double>::infinity();
    info.kappaB = (sB(2) > tol) ? sB(0) / sB(2) : std::numeric_limits<double>::infinity();

    int m = 3;
    bool typeI  = info.rankB < m;
    bool typeII = info.rankA < m;

    if (typeI && typeII) {
        info.type = "Type III (combined)";
    } else if (typeI) {
        info.type = "Type I (serial)";
    } else if (typeII) {
        info.type = "Type II (parallel)";
    } else {
        info.type = "Regular";
    }

    return info;
}

int main() {
    Matrix3d A, B;
    // Fill A and B here (e.g., using a function similar to compute_AB in Python)

    // Dummy example:
    A.setIdentity();
    B.setIdentity();

    SingularityInfo info = classifySingularity(A, B);

    std::cout << "Type: " << info.type << std::endl;
    std::cout << "det(A) = " << info.detA
              << ", det(B) = " << info.detB << std::endl;
    std::cout << "rank(A) = " << info.rankA
              << ", rank(B) = " << info.rankB << std::endl;
    std::cout << "kappa(A) = " << info.kappaA
              << ", kappa(B) = " << info.kappaB << std::endl;

    return 0;
}
      
